pkg/ruler/rulestore/bucket_client.go

```diff
@@ -0,0 +1,312 @@
+	if userBucket.IsObjNotFoundErr(err) {
+		level.Debug(b.logger).Log("msg", "rule group does not exist", "name", objectKey)
+		return nil, rules.ErrGroupNotFound
+	}
+
+	if err != nil {
+		return nil, err
+	}
+	defer func() { _ = reader.Close() }()
+
+	buf, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+
+	if rg == nil {
+		rg = &rules.RuleGroupDesc{}
+	} else {
+		rg.Reset()
+	}
+
+	err = proto.Unmarshal(buf, rg)
+	if err != nil {
+		return nil, err
+	}
+
+	return rg, nil
+}
+
+func (b *BucketRuleStore) ListAllUsers(ctx context.Context) ([]string, error) {
+	var users []string
+	err := b.bucket.Iter(ctx, rulePrefix, func(user string) error {
+		user = strings.TrimSuffix(strings.TrimPrefix(user, rulePrefix+delim), delim)
+		iterErr := tenant.ValidTenantID(user)
+
+		// Ignore but log invalid users in the backend. We don't want a single corrupted tenant to break
+		// rules for everyone else.
+		if iterErr != nil {
+			level.Error(b.logger).Log("msg", "invalid tenant id found in rule store", "err", iterErr, "user", user)
+		}
+
+		users = append(users, user)
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("unable to list users in rule store bucket, %w", err)
+	}
+
+	return users, nil
+}
+
+func (b *BucketRuleStore) ListAllRuleGroups(ctx context.Context) (map[string]rules.RuleGroupList, error) {
+	users, err := b.ListAllUsers(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	perUserRuleGroupListMap := map[string]rules.RuleGroupList{}
+
+	for _, user := range users {
+		namespaces, err := b.listNamespacesForUser(ctx, user)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, namespace := range namespaces {
+			userBucket := newRuleStoreUserBucketClient(user, b.bucket)
+			nsEncoded := generateRuleObjectKey(namespace, "")
+			err = userBucket.Iter(ctx, nsEncoded, func(group string) error {
+				group = strings.TrimSuffix(strings.TrimPrefix(group, nsEncoded), delim)
+				decodedGroup, err := base64.URLEncoding.DecodeString(group)
+				if err != nil {
+					return err
+				}
+				perUserRuleGroupListMap[user] = append(perUserRuleGroupListMap[user], &rules.RuleGroupDesc{
+					User:      user,
+					Namespace: namespace,
+					Name:      string(decodedGroup),
+				})
+				return nil
+			})
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return perUserRuleGroupListMap, nil
+}
+
+// ListRuleGroupsForUserAndNamespace returns all the active rule groups for a user from given namespace.
+// If namespace is empty, groups from all namespaces are returned.
+func (b *BucketRuleStore) ListRuleGroupsForUserAndNamespace(ctx context.Context, userID string, namespace string) (rules.RuleGroupList, error) {
+	userBucket := newRuleStoreUserBucketClient(userID, b.bucket)
+
+	var namespaces []string
+	var err error
+
+	if namespace != "" {
+		namespaces = []string{namespace}
+	} else {
+		namespaces, err = b.listNamespacesForUser(ctx, userID)
+		if err != nil {
+			return nil, err
```

Comment:
```suggestion
			return nil, errors.Wrapf(err, "failed to list namespaces for user %v", userID)
```

Reply:
👍 

I updated it to use `fmt.Errorf` with with the `%w` verb